README.md
[![GitHub version](https://badge.fury.io/gh/sm-react%2Freact-material-color-picker.svg)](https://badge.fury.io/gh/sm-react%2Freact-material-color-picker)
[![npm version](https://badge.fury.io/js/react-material-color-picker.svg)](https://badge.fury.io/js/react-material-color-picker)
[![@airbnb](https://img.shields.io/badge/code%20style-Airbnb-brightgreen.svg)](https://github.com/sm-react/react-material-color-picker/blob/master/.eslintrc)

# React Material Color Picker Component

[Material Design](https://material.google.com/) is a design language introduced by Google. If you want to find color inspiration for a specific design style based on [material color palette](https://material.google.com/style/color.html#), you can use this component as a development tool. You may find it useful while creating Material apps in combination with such libraries as [Material-UI](http://www.material-ui.com/)

## Appearance

![screen1](https://github.com/sm-react/react-material-color-picker/raw/master/doc/screen1.png)

![screen2](https://github.com/sm-react/react-material-color-picker/raw/master/doc/screen2.png)

![screen3](https://github.com/sm-react/react-material-color-picker/raw/master/doc/screen3.png)


## Install

$ npm i react-material-color-picker --save

## Usage

~~~
import React from 'react';
import MaterialColorPicker from 'react-material-color-picker';

// in your app
<MaterialColorPicker 
    initColor="rgba(0, 0, 0, 0.26)"
    onSubmit={actionLog()}
    onReset={actionLog()}
    style={{width: 400, backgroundColor: '#c7c7c7'}}
    submitLabel='Apply'
    resetLabel='Undo'
/>
~~~

## Demo
[![Live demo](https://img.shields.io/badge/Live%20Demo-%20Storybook-brightgreen.svg)](https://sm-react.github.io/react-material-color-picker/?selectedKind=Material%20Color%20Picker&selectedStory=default%20view&full=0&down=1&left=1&panelRight=1&downPanel=kadirahq%2Fstorybook-addon-actions%2Factions-panel)

## API

#### Props and Callbacks
**initColor** should be color string from Goggle material color palette

**style** - inline style of the root div node

**submitLabel** and **resetLabel** are titles of the appropriate buttons

**onSubmit** and **onReset** are callbacks wich will be invoked by clicking the appropriate buttons. It will recieve an argument with the following structure:
~~~js
event = {
        type, // 'submit' || 'reset', 
        timeStamp, // nativeEvent.timeStamp,
        target: {
            value, // currient color string,
            nativeEvent, // nativeEvent,
            name: 'MaterialColorPicker',
            node, // ref to root div element,
            ...this.props,
        }
};
~~~


## smArtLight
[![@UsulPro](https://img.shields.io/badge/github-UsulPro-blue.svg)](https://github.com/UsulPro)
[![@sm-react](https://img.shields.io/badge/github-smARTLight-red.svg)](https://github.com/sm-react)



CHANGELOG.md
## Changelog

### v1.0.0

* First stable release

props: initColor, onReset, onSubmit, resetLabel, style, submitLabel

### v1.1.0

* Added props: onSelect, onHover

src/tests/index.js
import React from 'react';
import { shallow, mount } from 'enzyme';
import { expect } from 'chai';
import sinon from 'sinon';
import MaterialColorPicker from '../MaterialColorPicker.jsx';
const { describe, it } = global;

describe('### <MaterialColorPicker /> ###\n\n', () => {
    it('should show 20 tone swatches', () => {
        const wrapper = shallow(<MaterialColorPicker />);
        const tonewrap = wrapper.find('.material-color-picker-tone-swatches');
        expect(tonewrap.children()).to.have.length(20);
    });

    it('should show right color name /* deepOrange800 */', () => {
        const wrapper = shallow(<MaterialColorPicker initColor="#d84315" />);
        const titlewrap = wrapper.find('.material-color-picker-title');
        expect(titlewrap.childAt(0).text()).to.be.equal('deepOrange800');
    });

    it('should show right color name /* darkBlack */', () => {
        const wrapper = shallow(<MaterialColorPicker initColor="rgba(0, 0, 0, 0.87)" />);
        const titlewrap = wrapper.find('.material-color-picker-title');
        expect(titlewrap.childAt(0).text()).to.be.equal('darkBlack');
    });

    it('should show right color in preview', () => {
        const initColor = 'rgba(255, 255, 255, 0.87)';
        const wrapper = shallow(<MaterialColorPicker initColor={initColor} />);
        const previewwrap = wrapper.find('.material-color-picker-preview');
        expect(previewwrap.props().style.backgroundColor)
            .to.be.equal(initColor);
    });

    it('should handle the onSubmit event', () => {
        const clickMe = sinon.stub();
    // Here we do a JSDOM render. So, that's why we need to
    // wrap this with a div.
        const wrapper = mount(
      <div>
        <MaterialColorPicker
          initColor="#40c4ff"
          onSubmit={ clickMe }
        />
      </div>
    );
        wrapper.find('.material-color-picker-submit').simulate('click');
        expect(clickMe.callCount).to.be.equal(1);
    });

    it('should switch colors tones', () => {
        const clickMe = sinon.stub();
        const wrapper = mount(
          <div>
            <MaterialColorPicker
              initColor="#f3e5f5"
              onSubmit={ clickMe }
            />
          </div>
        );

        const tonesw = wrapper.find('.material-color-picker-tone-swatches');
        const previewwrap = wrapper.find('.material-color-picker-preview');
        const satwrap = wrapper.find('.material-color-picker-sat-swatches');

        /* red */
        tonesw.childAt(0).simulate('click');
        expect(satwrap.children()).to.have.length(14);
        expect(previewwrap.props().style.backgroundColor)
            .to.be.equal('#ffebee');

        /* indigo */
        tonesw.childAt(4).simulate('click');
        expect(satwrap.children()).to.have.length(14);
        expect(previewwrap.props().style.backgroundColor)
            .to.be.equal('#e8eaf6');

        /* brown */
        tonesw.childAt(16).simulate('click');
        expect(satwrap.children()).to.have.length(10);
        expect(previewwrap.props().style.backgroundColor)
            .to.be.equal('#efebe9');

        /* grey */
        tonesw.childAt(18).simulate('click');
        expect(satwrap.children()).to.have.length(10);
        expect(previewwrap.props().style.backgroundColor)
            .to.be.equal('#fafafa');

        /* black */
        tonesw.childAt(19).simulate('click');
        expect(satwrap.children()).to.have.length(10);
        expect(previewwrap.props().style.backgroundColor)
            .to.be.equal(undefined);
    });
});



src/stories/index.js
import React from 'react';
import { storiesOf, action, setAddon } from '@kadira/storybook';
import infoAddon from '@kadira/react-storybook-addon-info';
import { withKnobs, text, object } from '@kadira/storybook-addon-knobs';

// import MaterialColorPicker from 'react-material-color-picker';
import MaterialColorPicker from '../MaterialColorPicker.jsx';


setAddon(infoAddon);


function actionLog(title) {
    return (event) => {
        const data = { ...event };
        data.target.node = 'console.log(event.target.node)';

        action(`${title || ''}${data.type}: ${data.target.value}`.trim())(data);
    };
}

storiesOf('Material Color Picker', module)
    .addDecorator((story) => (
        <div
          style={{
              position: 'relative',
              width: '100%',
//              minHeight: 600,
              overflowY: 'visible',
              display: 'flex',
              justifyContent: 'center',
          }}
        >
            {story()}
        </div>
    ))
    .addDecorator(withKnobs)
    .add('default view', () => (

        <div style={{ width: 400, backgroundColor: '#c7c7c7' }} >
            <MaterialColorPicker
              initColor="#3f51b5"
              onSubmit={actionLog()}
              onReset={actionLog()}
            />
        </div>

    ))
    .add('fullwidth view', () => (

        <div style={{ width: '70%' }} >
            <MaterialColorPicker
              initColor="#9ccc65"
              onSubmit={actionLog()}
              onReset={actionLog()}
              style={{ backgroundColor: '#c7c7c7' }}
            />
        </div>

    ))
    .add('events', () => (

        <div style={{ width: 400, backgroundColor: '#c7c7c7' }} >
            <MaterialColorPicker
              initColor="#3f51b5"
              onSubmit={actionLog()}
              onSelect={actionLog()}
              onHover={actionLog()}
              onReset={actionLog()}
            />
        </div>

    ))
//    storiesOf('Material Color Picker', module)
    .addWithInfo('API', `
[![GitHub version](https://badge.fury.io/gh/sm-react%2Freact-material-color-picker.svg)](https://badge.fury.io/gh/sm-react%2Freact-material-color-picker)
[![npm version](https://badge.fury.io/js/react-material-color-picker.svg)](https://badge.fury.io/js/react-material-color-picker)

#### Props and Callbacks
**initColor** should be color string from Goggle material color palette

**style** - inline style of the root div node

**submitLabel** and **resetLabel** are titles of the appropriate buttons

**onSubmit** and **onReset** are callbacks wich will be invoked by clicking the
appropriate buttons. It will recieve an argument with the following structure:
~~~js
event = {
        type, // 'submit' || 'reset',
        timeStamp, // nativeEvent.timeStamp,
        target: {
            value, // currient color string,
            nativeEvent, // nativeEvent,
            name: 'MaterialColorPicker',
            node, // ref to root div element,
            ...this.props,
        }
};
~~~

[![@UsulPro](https://img.shields.io/badge/github-UsulPro-blue.svg)](https://github.com/UsulPro)
[![@sm-react](https://img.shields.io/badge/github-smARTLight-red.svg)](https://github.com/sm-react)
`, () => (
            <div style={{ width: 600 }} >
               <MaterialColorPicker
                 initColor="rgba(0, 0, 0, 0.26)"
                 onSubmit={actionLog()}
                 onReset={actionLog()}
                 style={{ width: 400, backgroundColor: '#c7c7c7' }}
                 submitLabel="Apply"
                 resetLabel="Undo"
               />
            </div>
    ), { inline: true, source: false })
    .addWithInfo('try props',
`
->

->
#### select the \`KNOBS\` tab to get access for props
`,
                 () => {
                     const style = {
                         width: 400,
                         backgroundColor: '#c7c7c7',
                     };
                     return (<div >
            <MaterialColorPicker
              initColor={text('initColor', '#ec407a')}
              onSubmit={actionLog()}
              onReset={actionLog()}
              submitLabel={text('submitLabel', 'OK')}
              resetLabel={text('resetLabel', 'Cansel')}
              style={object('style', style)}
            />
        </div>);
                 }, { inline: true, source: false, header: false });
